### DISCUSSION TITLE
Possible to extract single cell-level gate memberships?

### DISCUSSION THREAD

**Discussion by hrj21** [DillonHammill/CytoExploreR#147] [Q&A] [answered]
Hi!

Thank you for creating such a powerful package! I know it's very easy to extract population-level statistics from a gatingSet, but is it possible to extract, at the per event level, which gates/nodes they belong to? A single event will often belong to multiple gates of course, so even if I was able to extract a logical vector per node, where the ith element is TRUE if the ith event belongs to node, and FALSE if not, that would be fine. Perhaps I'm overthinking and there is already an easy way to get something like this. Any thoughts you can offer would be much appreciated.

Best wishes
Hefin

**Comment by DillonHammill** [comment, marked as answer]
@hrj21, there isn't a nice CytoExploreR wrapper for this as of yet but you can have a look at `gh_pop_get_indices()` and `gh_pop_get_indices_mat()` from `flowWorkspace`. I will add a wrapper for the latter function which will return a logical vector for multiple populations at once. Here is an example using the `Activation` dataset shipped with `CytoExploreRData`:
```{r}
library(CytoExploreRData)
gs <- GatingSet(Activation)
gs <- cyto_compensate(gs)
gs <- cyto_transform(gs)
gs <- cyto_gatingTemplate_apply(gs, Activation_gatingTemplate)

gh_pop_get_indices_mat(gs[[32]], c("T Cells", "CD4 T Cells", "CD8 T Cells"))
```
These APIs operate at the level of the `GatingHierarchy` so you will need to loop through the samples in your `GatingSet` to get the indices for every sample - I will write a wrapper to automate this within `CytoExploreR`.

I have just added a new `cyto_gate_indices()` function to version 2.0.0 of CytoExploreR (coming soon) that will automatically pull out the indices for every sample:
```{r}
cyto_gate_indices(gs, select = 31:32, c("T Cells", "CD4 T Cells", "CD8 T Cells"))
```

**Reply by hrj21** [reply]
Outstanding, thank you so much! I ended up doing this to get a matrix for all nodes in the gs:
`gh_pop_get_indices_mat(gs[[1]], basename(cyto_nodes(gs)))`

Perhaps there's a better way of doing it but that achieved what I wanted. Thanks again for your help and for this excellent package.
